frc46_token/src/token/state.rs

```diff
@@ -506,114 +511,169 @@ impl TokenState {
+                    if balance.is_zero() {
+                        errors.push(StateInvariantError::ExplicitZeroBalance(owner));
+                    }
+
+                    // balances should be a multiple of granularity
+                    let (_, modulus) = balance.div_rem(granularity);
+                    if !modulus.is_zero() {
+                        errors.push(StateInvariantError::InvalidGranularity {
+                            balance: balance.clone(),
+                            owner,
+                            granularity,
+                        });
+                    }
+
+                    // track total balance
+                    balance_sum = balance_sum.clone() + balance.clone();
+
+                    // clone into HashMap
+                    balance_map.insert(owner, balance.clone());
+                }
+                Ok(())
+            })
+            .unwrap();
 
         // all balances must add up to total supply
         if balance_sum.ne(&self.supply) {
-            return Err(StateInvariantError::BalanceSupplyMismatch {
+            errors.push(StateInvariantError::BalanceSupplyMismatch {
                 supply: self.supply.clone(),
                 balance_sum,
             });
         }
 
-        let mut maybe_err: Option<StateInvariantError> = None;
         // check allowances are all non-negative
-        let allowances_map = self.get_allowances_map(bs)?;
-        let res = allowances_map.for_each(|owner, _| {
-            let owner = match decode_actor_id(owner) {
-                None => {
-                    bail!("invalid owner key in allowances map")
-                }
-                Some(a) => a,
-            };
-            let allowance_map = self.get_owner_allowance_map(bs, owner)?;
-            // check that the allowance map isn't empty
-            if allowance_map.is_none() {
-                maybe_err = Some(StateInvariantError::ExplicitEmptyAllowance(owner));
-                bail!("invariant failed")
+        let allowances_hamt = match self.get_allowances_map(bs) {
+            Ok(map) => map,
+            Err(e) => {
+                errors.push(StateInvariantError::State(e));
+                return Err(errors);
             }
+        };
 
-            let allowance_map = allowance_map.unwrap();
-            allowance_map.for_each(|operator, allowance| {
-                let operator = match decode_actor_id(operator) {
-                    None => {
-                        bail!("invalid operator key in allowances map")
+        let mut global_allowances_map: HashMap<ActorID, BTreeMap<ActorID, TokenAmount>> =
+            HashMap::new();
+        allowances_hamt
+            .for_each(|owner, _| {
+                if let Some(owner) = decode_key_addr(owner, &mut errors) {
+                    let allowance_map = self.get_owner_allowance_map(bs, owner)?;
+
+                    // check that the allowance map exists
+                    if allowance_map.is_none() {
```

Comment:
I don't think this can happen, since we're in for_each. Would be more obvious if you used the cid value rather than doing a get() for it again.

Reply:
The intent here is to verify that the allowance map can be looked up by the owner correctly (i.e. that the cid in the global map is not invalid or stale). I've added an explicit check that the cids should match.